Formula/nginx-unit.rb

```diff
@@ -0,0 +1,267 @@
+    <plist version="1.0">
+      <dict>
+        <key>Label</key>
+        <string>#{plist_name}</string>
+        <key>RunAtLoad</key>
+        <true/>
+        <key>KeepAlive</key>
+        <false/>
+        <key>ProgramArguments</key>
+        <array>
+            <string>#{opt_bin}/unitd</string>
+            <string>--no-daemon</string>
+        </array>
+        <key>WorkingDirectory</key>
+        <string>#{HOMEBREW_PREFIX}</string>
+      </dict>
+    </plist>
+    EOS
+  end
+
+  test do
+    system "#{bin}/unitd", "--version"
+
+    require "json"
+    require "net/http"
+    require "socket"
+    require "timeout"
+
+    logfile = var/"log/unit.log"
+    logfile.unlink if logfile.exist?
+
+    pidfile = var/"run/unit.pid"
+    pidfile.unlink if pidfile.exist?
+
+    begin
+      # We need to install rack for the ruby module test to work, so set a
+      # temporary gems path before booting unitd. (unitd boots and functions
+      # fine without, but fails to start an app until rack is installed.)
+      rubydir = testpath/"ruby"
+      rubygems = rubydir/"gems"
+      rubygems.mkpath
+      ENV["GEM_PATH"] = rubygems.to_s
+      system "gem", "install", "rack", "--install-dir=#{rubygems}", "--verbose"
+
+      pid = spawn "#{bin}/unitd", "--no-daemon"
+
+      timeout(5) { sleep 0.1 until pidfile.exist? }
+
+      assert_equal pid, pidfile.read.chomp.to_i
+
+      timeout(5) { sleep 0.1 until File.exist? var/"run/unit.control.sock" }
+
+      # Unit uses a unix control socket. Net::HTTP to a socket is hard. Sorry.
+      socket = Net::BufferedIO.new(UNIXSocket.new(var/"run/unit.control.sock"))
+      request = Net::HTTP::Get.new("/")
+      request.exec(socket, "1.1", request.path)
+      response = Net::HTTPResponse.read_new(socket)
+      response.reading_body(socket, request.response_body_permitted?) { }
+
+      assert_equal("200", response.code)
+      expected_body = {"listeners" => {}, "applications" => {}}
+      assert_equal expected_body, JSON.parse(response.body)
+
+      assert logfile.size > 0
+
+      # Unit should discover each of the modules and talk about them in the log
+      logfile.read.tap do |logs|
+        #assert_include logs, "lib/unit/go.unit.so"
+        assert_include logs, "lib/unit/perl.unit.so"
+        #assert_include logs, "lib/unit/php.unit.so"
+        assert_include logs, "lib/unit/python.unit.so"
+        assert_include logs, "lib/unit/ruby.unit.so"
+      end
+
+      ### Perl module
+
+      socket = Net::BufferedIO.new(UNIXSocket.new(var/"run/unit.control.sock"))
+      request = Net::HTTP::Put.new("/config")
+      request.body = JSON.generate({})
+      request.exec(socket, "1.1", request.path)
+      response = Net::HTTPResponse.read_new(socket)
+      response.reading_body(socket, request.response_body_permitted?) { }
+
+      perldir = testpath/"perl"
+      perldir.mkpath
+      perlapp = perldir/"app.psgi"
+
+      perlapp.write <<~PERL
```

Comment:
Standardise on `EOS`, but you can drop the `mkpath` above as the `.write` here will create the file as necessary.

Reply:
I have syntax highlighting for heredocs based no the tag name, which is nice, but I'll change them to EOS.

I didn't know `#write` did `#mkpath` — neat! TIL.